Library/Homebrew/cmd/install.rb

```diff
@@ -135,13 +135,45 @@ def install
           raise "No devel block is defined for #{f.full_name}"
         end
 
-        current = f if f.installed?
-        current ||= f.old_installed_formulae.first
+        installed_head_version = f.latest_head_version
+        new_head_installed = installed_head_version &&
+          !f.head_version_outdated?(installed_head_version, fetch_head: ARGV.fetch_head?)
 
-        if current
-          msg = "#{current.full_name}-#{current.installed_version} already installed"
-          unless current.linked_keg.symlink? || current.keg_only?
-            msg << ", it's just not linked"
+        prefix_installed = (f.prefix.exist? && !f.prefix.children.empty?)
+
+        if f.keg_only? && f.any_version_installed? && f.optlinked? && !ARGV.force?
+          # keg-only install is only possible when no other version is
+          # linked to opt, because installing without any warnings can break
+          # dependencies. Therefore before performing other checks we need to be
+          # sure --force flag is passed.
+          opoo "#{f.full_name} is a keg-only and another version is linked to opt."
+          puts "Use `brew install --force` if you want to install this version"
+        elsif ARGV.build_head? && new_head_installed || prefix_installed
+          # After we're sure that --force flag is passed for linked to opt
+          # keg-only we need to be sure that the version we're attempting to
+          # install is not already installed.
+
+          installed_version = if ARGV.build_head?
+            f.latest_head_version
+          else
+            f.pkg_version
+          end
+
+          msg = "#{f.full_name}-#{installed_version} already installed"
+          if f.linked? && f.linked_version != installed_version
```

Comment:
Could make `f.linked_version != installed_version` another variable for clarity but feel free to ignore that.